traversal/traversal/traversal.py

```diff
@@ -0,0 +1,211 @@
+from pcconfig import config
+
+import pynecone as pc
+import random
+import asyncio
+from collections import deque
+
+GRID_SIZE = 7
+
+def to_matrix(l, n):
+    """Convert a 1D array to a 2D array."""
+    return [l[i : i + n] for i in range(0, len(l), n)]
+
+def generate_graph(walls, size) -> list[list[int]]:
+    """Generate a 2D grid of size x size with walls number of walls."""
+    color = [0] * (size**2)
+
+    while walls > 0:
+        index = random.randint(size, size**2 - 1)
+        if color[index] != "blue":
+            color[index] = "blue"
+            walls -= 1
+
+    color[random.randint(0, size - 1)] = "red"
+    color[random.randint(size, size**2 - 1)] = "green"
+
+    return to_matrix(color, GRID_SIZE)
+
+
+class State(pc.State):
+    """The app state."""
+
+    option: str = ""
+    walls: int = 0
+    colored_graph: list[list[int]] = generate_graph(walls, GRID_SIZE)
+    initial: bool = True
+    s: list = []
+    q: list = []
+
+    def set_walls(self, value):
+        """Set the number of walls."""
+        if value != "":
+            if int(value) >= 0:
+                self.walls = int(value)
+
+    def new_graph(self):
+        """Reset the state with parameters."""
+        self.colored_graph = generate_graph(self.walls, GRID_SIZE)
+        self.initial = True
+        self.s = []
+        self.q = []
+
+    def run(self):
+        """Run the selected algorithm."""
+        if self.option == "DFS":
+            return self.dfs()
+        elif self.option == "BFS":
+            return self.bfs()
+
+    def dfs(self):
+        return self.run_dfs
+
+    async def run_dfs(self):
+        await asyncio.sleep(0.01)
+        colors = self.colored_graph
+
+        if self.initial:
+            for i in range(len(colors)):
+                for j in range(len(colors[i])):
+                    if colors[i][j] == "red":
+                        self.s.append((i, j))
+                        self.initial = False
+                        break
+
+        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
+
+        if self.s:
+            i, j = self.s.pop()
+
+            if colors[i][j] == "green":
+                self.colored_graph = colors
+                return
+
+            if colors[i][j] != "red":
+                colors[i][j] = "yellow"
+
+            self.colored_graph = colors
+
+            for di, dj in directions:
+                i2, j2 = i + di, j + dj
+                if (
+                    0 <= i2 < len(colors)
+                    and 0 <= j2 < len(colors[i2])
+                    and colors[i2][j2] != "yellow"
+                    and colors[i2][j2] != "blue"
+                ):
+                    self.s.append((i2, j2))
+            return self.run_dfs
+
+        return pc.window_alert("No path found")
+
+    def bfs(self):
+        return self.run_bfs
+
+    async def run_bfs(self):
+        await asyncio.sleep(0.0000000000000001)
```

Comment:
Should we sleep for 0.01 like we do in the `dfs`? Otherwise may not be able to see the animation (though I guess the websocket latency will add its own time lol)